examples/naive/aarch64/intt_dilithium_1234_5678_manual_ld4.s

```diff
@@ -0,0 +1,512 @@
+// The macros are not unfolded by Slothy and thus interpreted as instructions,
+// which are easier to parse due to e.g. the lack of size specifiers and simpler
+// syntax for pre and post increment for loads and stores.
+//
+// Eventually, NeLight should include a proper parser for AArch64,
+// but for initial investigations, the below is enough.
+
+.macro ldr_vo vec, base, offset
+        ldr qform_\vec, [\base, #\offset]
+.endm
+.macro ldr_vi vec, base, inc
+        ldr qform_\vec, [\base], #\inc
+.endm
+.macro str_vo vec, base, offset
+        str qform_\vec, [\base, #\offset]
+.endm
+.macro str_vi vec, base, inc
+        str qform_\vec, [\base], #\inc
+.endm
+.macro vsub d,a,b
+        sub \d\().4s, \a\().4s, \b\().4s
+.endm
+.macro vadd d,a,b
+        add \d\().4s, \a\().4s, \b\().4s
+.endm
+.macro vqrdmulh d,a,b
+        sqrdmulh \d\().4s, \a\().4s, \b\().4s
+.endm
+.macro vmul d,a,b
+        mul \d\().4s, \a\().4s, \b\().4s
+.endm
+.macro vmls d,a,b
+        mls \d\().4s, \a\().4s, \b\().4s
+.endm
+.macro vqrdmulhq d,a,b,i
+        sqrdmulh \d\().4s, \a\().4s, \b\().s[\i]
+.endm
+.macro vmulq d,a,b,i
+        mul \d\().4s, \a\().4s, \b\().s[\i]
+.endm
+.macro vmlsq d,a,b,i
+        mls \d\().4s, \a\().4s, \b\().s[\i]
+.endm
+.macro trn1_d d,a,b
+        trn1 \d\().2d, \a\().2d, \b\().2d
+.endm
+.macro trn2_d d,a,b
+        trn2 \d\().2d, \a\().2d, \b\().2d
+.endm
+.macro trn1_s d,a,b
+        trn1 \d\().4s, \a\().4s, \b\().4s
+.endm
+.macro trn2_s d,a,b
+        trn2 \d\().4s, \a\().4s, \b\().4s
+.endm
+
+.macro mulmodq dst, src, const, idx0, idx1
+        vmulq       \dst,  \src, \const, \idx0
+        vqrdmulhq   \src,  \src, \const, \idx1
+        vmls        \dst,  \src, modulus
+.endm
+
+.macro mulmod dst, src, const, const_twisted
+        vmul       \dst,  \src, \const
+        vqrdmulh   \src,  \src, \const_twisted
+        vmls       \dst,  \src, modulus
+.endm
+
+.macro montg_reduce a
+        srshr tmp.4S,  \a\().4S, #23
+        vmls   \a, tmp, modulus
+.endm
+
+.macro canonical_reduce a, modulus_half, neg_modulus_half, tmp1, tmp2
+        cmge \tmp1\().4s, \neg_modulus_half\().4s, \a\().4s
+        cmge \tmp2\().4s, \a\().4s, \modulus_half\().4s
+        sub \tmp2\().4s, \tmp1\().4s, \tmp2\().4s
+        vmls \a, \tmp2, modulus
+.endm
+
+.macro gs_butterfly a, b, root, idx0, idx1
+        vsub     tmp,    \a, \b
+        vadd     \a,    \a, \b
+        mulmodq  \b, tmp, \root, \idx0, \idx1
+.endm
+
+.macro mulmod_v dst, src, const, const_twisted
+        vmul        \dst,  \src, \const
+        vqrdmulh    \src,  \src, \const_twisted
+        vmls        \dst,  \src, modulus
+.endm
+
+.macro gs_butterfly_v a, b, root, root_twisted
+        vsub    tmp,    \a, \b
+        vadd    \a,    \a, \b
+        mulmod  \b, tmp, \root, \root_twisted
+.endm
+
+.macro mul_ninv dst0, dst1, dst2, dst3, dst4, dst5, dst6, dst7, src0, src1, src2, src3, src4, src5, src6, src7
```

Comment:
This is an optional improvement, but it may be better to address it while we're at it:

Ideally `mulmod` uses a temporary (perhaps fixed throughout the clean code) for intermediates, so that it works regardless of whether `a` and `b` alias. This would allow you to pass just in=out to `mul_ninv`. Things like overwriting `data0,..,data3` with the `mul_ninv` result of `datat4..7`, and thus effectively using `data0..3` as temporaries, should ideally not be present in clean code.

Reply:
I agree, changed the way the macro works for the {inv}NTTs.